src/rgw/rgw_op.cc

```diff
@@ -937,37 +937,94 @@ void handle_replication_status_header(
+        if (sync_cloudtiered) {
           bufferlist t, t_tier;
-          RGWObjTier tier_config;
-          m.get_tier_config(&tier_config);
-
           t.append("cloud-s3");
           attrs[RGW_ATTR_CLOUD_TIER_TYPE] = t;
           encode(tier_config, t_tier);
           attrs[RGW_ATTR_CLOUD_TIER_CONFIG] = t_tier;
+          return op_ret;
         }
+        if (tier_config.tier_placement.allow_read_through) {
+          uint64_t restore_days = tier_config.tier_placement.read_through_restore_days;
+          attr_iter = attrs.find(RGW_ATTR_RESTORE_STATUS);
+          rgw::sal::RGWRestoreStatus restore_status;
+          if (attr_iter != attrs.end()) {
+            bufferlist bl = attr_iter->second;
+            auto iter = bl.cbegin();
+            decode(restore_status, iter);
+          }
+          if (attr_iter == attrs.end() || restore_status == rgw::sal::RGWRestoreStatus::RestoreFailed) {
+            rgw::sal::Bucket* pbucket = NULL;
+            pbucket = s->bucket.get();
+            std::unique_ptr<rgw::sal::PlacementTier> tier;
+            rgw_placement_rule target_placement;
+            target_placement.inherit_from(pbucket->get_placement_rule());
+            attr_iter = attrs.find(RGW_ATTR_STORAGE_CLASS);
+            if (attr_iter != attrs.end()) {
+              target_placement.storage_class = attr_iter->second.to_str();
+            }
+            op_ret = driver->get_zone()->get_zonegroup().get_placement_tier(target_placement, &tier);
+            ldpp_dout(dpp, 20) << "getting tier placement handle cloud tier" << op_ret <<
+                       " storage class " << target_placement.storage_class << dendl;
+            if (op_ret < 0) {
+              return op_ret;
+            }
+            rgw_bucket_dir_entry ent;
+            ent.key.name = s->object->get_key().name;
+            ent.meta.accounted_size = ent.meta.size = s->obj_size;
+            ent.meta.etag = "" ;
+            ceph::real_time mtime = s->object->get_mtime();
+            uint64_t epoch = 0;
+            op_ret = get_system_versioning_params(s, &epoch, NULL);
+            ldpp_dout(dpp, 20) << "getting versioning params tier placement handle cloud tier" << op_ret << dendl;
+            // TODO : pass proper values for etag, days and fill up necessary fields in ent
+            op_ret = s->object->restore_obj_from_cloud(pbucket, tier.get(), target_placement, ent, s->cct, tier_config,
+                                                   mtime, epoch, restore_days, dpp, y, s->bucket->get_info().flags);
+            if (op_ret < 0) {
+              ldpp_dout(dpp, 0) << "object " << ent.key.name << " fetching failed" << op_ret << dendl;
+              s->err.message = "failed to restore object";
+              return op_ret;
+            }
+            ldpp_dout(dpp, 20) << "object " << ent.key.name << " fetching succeed" << dendl;
+          }
+          if (attr_iter == attrs.end() || restore_status != rgw::sal::RGWRestoreStatus::CloudRestored) {
```

Comment:
Additional check needs to be added to see if the status if InProgress. As per the spec,

To get the status of object restoration, you can send a HEAD request. Operations return the x-amz-restore header, which provides information about the restoration status, in the response. You can use Amazon S3 event notifications to notify you when a restore is initiated or completed. For more information, see [Configuring Amazon S3 Event Notifications](https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html) in the Amazon S3 User Guide.

So we need to set x-amz-restore  header as InProgress in such cases.

Reply:
This probably can be addressed in another patch.. please add  this as TODO .